### PR TITLE
Support optional id_tokens in refresh responses

### PR DESCRIPTION
**Pull request by howzat**
## Allowing refresh requests to succeed when no id_token is provided in the response
This PR makes the following changes
- When refreshing the access_token if no id_token is supplied the access_token will still be refreshed and updated in the session. If an id_token is also provided then the session is also populated with that information (presuming the token is successfully validated)
- The expiry time of the session is now taken from the **outh** token expiry and _not_ the id_token. It is not uncommon for access tokens have a relatively short expiry time, since they can be extended using the refresh. Using the expiry time as defined on the id_token leads to a situation where access tokens are expiring long before the id_token. The session is then effectively broken since it contains an expired access_token that will not be refreshed until such time as the id_token expires.
- The code has been refactored to allow for better unit test coverage
- The major paths of Redeeming and Refreshing the access token have some test coverage

## Motivation and Context
This change is to address the bug raised https://github.com/pusher/oauth2_proxy/issues/318

## How Has This Been Tested?
The tests that exercise the Redeem and RefreshTokenIfNeeded paths have also been run against the current master _oidc.go_  and then against this branch to verify the red/green  behaviour changes are happening as expected.

I've run the oauth2_proxy locally using an Okta App/Authorisation server. I've been able to see the Redeem and Refresh paths working for token responses that _do_ include id_token. I've also deployed my own fork into our context and observed it working when no id token is present in the refresh response.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by howzat** [comment]
@JoelSpeed I've made some progress with fixing https://github.com/pusher/oauth2_proxy/issues/318

My intention is to continue to extend the unit tests for the oidc provider, but before I go any further I'd appreciate a small review on the work that I've done. You can tell me if you see value in me continuing. 

A couple of points that are worth highlighting early...
- Do you agree with my understanding that the primary purpose of the refresh (and refresh_token) is to obtain a new access token? And that whilst that mechanism can also be used to update an id_token, that's basically a vendor specific operation. 
- I'm an experienced programmer, but I've never written any Golang before (I'm sure you can tell!). I'm very comfortable taking criticism and happy to learn. So please set me straight if you see things that offend proper Go programmers :)

**Comment by syscll** [comment]
@JoelSpeed @steakunderscore this looks like an issue we should prioritise.

File: providers/oidc.go [hunk 1]
```diff
@@ -84,55 +92,129 @@ func (p *OIDCProvider) redeemRefreshToken(s *sessions.SessionState) (err error)
...
+		return p.Verifier.Verify(ctx, rawIdToken)
 	}
 
-	// Parse and verify ID Token payload.
-	idToken, err := p.Verifier.Verify(ctx, rawIDToken)
+	idToken, err := p.extractIDToken(getIDToken, verifyIDToken)
+	return idToken, err
+}
+
+func (p *OIDCProvider) extractIDToken(findIDToken GetIDTokenFn, verifyIDToken VerifyIDTokenFn) (*oidc.IDToken, error) {
```
[the remark this answers is not in the document]
Comment (howzat) [reply providers/oidc.go]: Yes, it was to make it easier to test. What you suggest makes sense, I'll give it a go.

File: providers/oidc.go [hunk 2]
```diff
@@ -84,55 +92,129 @@ func (p *OIDCProvider) redeemRefreshToken(s *sessions.SessionState) (err error)
...
+			return nil, fmt.Errorf("couldn't extract claims from id_token (%e)", err)
+		}
+
+		if claims != nil {
+
+			if !p.AllowUnverifiedEmail && claims.Verified != nil && !*claims.Verified {
+				return nil, fmt.Errorf("email in id_token (%s) isn't verified", claims.Email)
+			}
+
+			newSession.IDToken = token.Extra("id_token").(string) // can this be derived from the object version
```
[the remark this answers is not in the document]
Comment (howzat) [reply providers/oidc.go]: I've got to be honest, I can't remember. I'll remove it unless it comes back to me

**Comment by howzat** [comment]
@JoelSpeed I've updated (hopefully improved) the code in line with your suggestion. The tests are also a little neater and easier to reason about too.

**Pull request merged**